include/nbl/builtin/hlsl/math/quaternions.hlsl

```diff
@@ -242,44 +287,63 @@ struct quaternion
         return precompPart * cosAngle + hlsl::cross(planeNormal, precompPart);
     }
 
-    this_t inverse()
+    this_t inverse() NBL_CONST_MEMBER_FUNC
     {
         this_t retval;
-        retval.data.x = bit_cast<scalar_type>(bit_cast<AsUint>(data.x)^0x80000000u);
-        retval.data.y = bit_cast<scalar_type>(bit_cast<AsUint>(data.y)^0x80000000u);
-        retval.data.z = bit_cast<scalar_type>(bit_cast<AsUint>(data.z)^0x80000000u);
+        retval.data.xyz = -retval.data.xyz;
         retval.data.w = data.w;
         return retval;
     }
 
-    static this_t normalize(NBL_CONST_REF_ARG(this_t) q)
+    data_type data;
+};
+
+}
+
+
+namespace cpp_compat_intrinsics_impl
+{
+template<typename T>
+struct normalize_helper<math::truncated_quaternion<T> >
+{
+    static inline math::truncated_quaternion<T> __call(const math::truncated_quaternion<T> q)
     {
-        this_t retval;
+        math::truncated_quaternion<T> retval;
         retval.data = hlsl::normalize(q.data);
```

Comment:
truncated quaternion is normalized by definition (last component is dropped), this should be a NOOP